File: backend/Shared/EventPlanr.Application/Models/Pagination/PaginatedListDto.cs
﻿namespace EventPlanr.Application.Models.Pagination;

public class PaginatedListDto<T>
{
    public List<T> Items { get; set; }
    public int PageNumber { get; set; }
    public int TotalPages { get; set; }
    public int TotalCount { get; set; }

    public PaginatedListDto(List<T> items, int count, int pageNumber, int pageSize)
    {
        PageNumber = pageNumber;
        TotalPages = (int)Math.Ceiling(count / (double)pageSize);
        TotalCount = count;
        Items = items;
    }

    public bool HasPreviousPage => PageNumber > 1;

    public bool HasNextPage => PageNumber < TotalPages;
}

File: frontend/lib/data/network/models/event_planr/common/paginated_list_dto.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'paginated_list_dto.freezed.dart';
part 'paginated_list_dto.g.dart';

@freezed
class PaginatedListDto with _$PaginatedListDto {
  const factory PaginatedListDto({
    required List<Map<String, dynamic>> items,
    required int pageNumber,
    required int totalPages,
    required int totalCount,
    required bool hasPreviousPage,
    required bool hasNextPage,
}) = _PaginatedListDto;

  factory PaginatedListDto.fromJson(Map<String, Object?> json) =>
      _$PaginatedListDtoFromJson(json);
}
